File: packages/mvn/src/main/kotlin/app/lacabra/sailor/responses/DomainGetResponse.kt
package app.lacabra.sailor.responses

import app.lacabra.sailor.DomainCategory
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The response from the domain get request
 * if the domain was not found, the response will be null
 */
@Serializable
data class DomainGetResponse(
    /**
     * The domain that was checked
     */
    val domain: String,
    /**
     * The category of the domain
     */
    val category: DomainCategory,
    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    val severity: Int,
    /**
     * The public notes of the domain left by the moderators
     */
    val notes: String,
    /**
     * The user that submitted the domain
     */
    @SerialName("submitted_by")
    val submittedBy: String,
    /**
     * The date the domain was submitted
     */
    @SerialName("submitted_at")
    val submittedAt: String,
)

/**
 * The response from the domain get request with a key
 * if the domain was not found, the response will be null
 */
@Serializable
data class DomainGetAdvancedResponse(
    /**
     * The internal ID of the domain
     */
    val id: Int,

    /**
     * The domain that was checked
     */
    val domain: String,

    /**
     * The category of the domain
     */
    val category: DomainCategory,

    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    val severity: Int,

    /**
     * The public notes of the domain left by the moderators
     */
    @SerialName("public_notes")
    val publicNotes: String,

    /**
     * The user that submitted the domain
     */
    @SerialName("submitted_by")
    val submittedBy: String,

    /**
     * The date the domain was submitted
     */
    @SerialName("submitted_at")
    val submittedAt: String,

    /**
     * The reason the user gave when submitting the domain
     */
    @SerialName("submitted_reason")
    val submittedReason: String,

    /**
     * The moderator that approved the domain
     */
    @SerialName("approved_by")
    val approvedBy: String,

    /**
     * The date the domain was approved
     */
    @SerialName("approved_at")
    val approvedAt: String,

    /**
     * The note the moderator left when approving the domain
     */
    val notes: String,

    /**
     * How many times the domain has been consulted
     */
    @SerialName("times_consulted")
    val timesConsulted: Int,
)

/**
 * The response from the domain get request with a key
 * if the domain was not found, the response will be null
 */
@Serializable
data class DomainGetMasterResponse(
    /**
     * The internal ID of the domain
     */
    val id: Int,

    /**
     * The domain that was checked
     */
    val domain: String,

    /**
     * The category of the domain
     */
    val category: DomainCategory,

    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    val severity: Int,

    /**
     * The public notes of the domain left by the moderators
     */
    @SerialName("public_notes")
    val publicNotes: String,

    /**
     * The user that submitted the domain
     */
    @SerialName("submitted_by")
    val submittedBy: String,

    /**
     * The date the domain was submitted
     */
    @SerialName("submitted_at")
    val submittedAt: String,

    /**
     * The IP address of the user that submitted the domain, only if the key is a master key
     */
    @SerialName("submitted_ip")
    val submittedIp: String,

    /**
     * The user agent of the user that submitted the domain, only if the key is a master key
     */
    @SerialName("submitted_user_agent")
    val submittedUserAgent: String,

    /**
     * The reason the user gave when submitting the domain
     */
    @SerialName("submitted_reason")
    val submittedReason: String,

    /**
     * The moderator that approved the domain
     */
    @SerialName("approved_by")
    val approvedBy: String,

    /**
     * The date the domain was approved
     */
    @SerialName("approved_at")
    val approvedAt: String,

    /**
     * The key that was used to approve the domain, only if the key is a master key
     */
    @SerialName("approved_key")
    val approvedKey: String,

    /**
     * The note the moderator left when approving the domain
     */
    val notes: String,

    /**
     * How many times the domain has been consulted
     */
    @SerialName("times_consulted")
    val timesConsulted: Int,
)

File: packages/npm/src/responses/domainGetResponse.ts
export type DomainGetResponseRaw = {
    /**
     * The domain that was checked
     */
    domain: string,
    /**
     * The category of the domain
     */
    category: string,
    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    severity: number,
    /**
     * The public notes of the domain left by the moderators
     */
    notes: string,
    /**
     * The user that submitted the domain
     */
    submitted_by: string,
    /**
     * The date the domain was submitted
     */
    submitted_at: string
}

export type DomainGetResponse =
    Omit<DomainGetResponseRaw, "submitted_at">
    & Omit<DomainGetResponseRaw, "submitted_by">
    & { submittedBy: string, submittedAt: string }

export type DomainGetAdvancedResponseRaw = {
    /**
     * The internal ID of the domain
     */
    id: number,
    /**
     * The domain that was checked
     */
    domain: string,
    /**
     * The category of the domain
     */
    category: string,
    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    severity: number,
    /**
     * The note the moderator left when approving the domain
     */
    notes: string,
    /**
     * The user that submitted the domain
     */
    submitted_by: string,
    /**
     * The date the domain was submitted
     */
    submitted_at: string,
    /**
     * The public notes of the domain left by the moderators
     */
    public_notes: string,
    /**
     * The reason the user gave when submitting the domain
     */
    submitted_reason: string,
    /**
     * The moderator that approved the domain
     */
    approved_by: string,
    /**
     * How many times the domain has been consulted
     */
    times_consulted: number
}

export type DomainGetAdvancedResponse =
    Omit<DomainGetAdvancedResponseRaw, "submitted_at">
    & Omit<DomainGetAdvancedResponseRaw, "submitted_by">
    & Omit<DomainGetAdvancedResponseRaw, "times_consulted">
    & Omit<DomainGetAdvancedResponseRaw, "approved_by">
    & Omit<DomainGetAdvancedResponseRaw, "submitted_reason">
    & Omit<DomainGetAdvancedResponseRaw, "public_notes">
    & {
    submittedBy: string,
    submittedAt: string,
    timesConsulted: number,
    approvedBy: string,
    submittedReason: string,
    publicNotes: string
}


export type DomainGetMasterResponseRaw = {
    /**
     * The internal ID of the domain
     */
    id: number,
    /**
     * The domain that was checked
     */
    domain: string,
    /**
     * The category of the domain
     */
    category: string,
    /**
     * The severity of the domain. 0 is the lowest and 10 is the highest depending on how dangerous the domain is.
     * Example: How easy it is for a user to be tricked into giving away their credentials.
     */
    severity: number,
    /**
     * The note the moderator left when approving the domain
     */
    notes: string,
    /**
     * The user that submitted the domain
     */
    submitted_by: string,
    /**
     * The date the domain was submitted
     */
    submitted_at: string,
    /**
     * The public notes of the domain left by the moderators
     */
    public_notes: string,
    /**
     * The reason the user gave when submitting the domain
     */
    submitted_reason: string,
    /**
     * The moderator that approved the domain
     */
    approved_by: string,
    /**
     * How many times the domain has been consulted
     */
    times_consulted: number,
    /**
     * The IP address of the user that submitted the domain, only if the key is a master key
     */
    submitted_ip: string,
    /**
     * The user agent of the user that submitted the domain, only if the key is a master key
     */
    submitted_user_agent: string,
    /**
     * The key that was used to approve the domain, only if the key is a master key
     */
    approved_key: string,
}

export type DomainGetMasterResponse =
    Omit<DomainGetMasterResponseRaw, "submitted_at">
    & Omit<DomainGetMasterResponseRaw, "submitted_by">
    & Omit<DomainGetMasterResponseRaw, "times_consulted">
    & Omit<DomainGetMasterResponseRaw, "approved_by">
    & Omit<DomainGetMasterResponseRaw, "submitted_reason">
    & Omit<DomainGetMasterResponseRaw, "public_notes">
    & Omit<DomainGetMasterResponseRaw, "approved_key">
    & Omit<DomainGetMasterResponseRaw, "submitted_user_agent">
    & Omit<DomainGetMasterResponseRaw, "submitted_ip">
    & {
    submittedBy: string,
    submittedAt: string,
    timesConsulted: number,
    approvedBy: string,
    submittedReason: string,
    publicNotes: string,
    approvedKey: string,
    submittedUserAgent: string,
    submittedIp: string
}